cardano-api/test/Test/Cardano/Api/TxBody.hs

```diff
@@ -0,0 +1,285 @@
+  TxWithdrawals _ [] -> TxWithdrawalsNone
+  other              -> other
+
+-- | Unify empty and None.
+normalizeCertificates :: TxCertificates ViewTx era -> TxCertificates ViewTx era
+normalizeCertificates = \case
+  TxCertificates _ [] _ -> TxCertificatesNone
+  other                 -> other
+
+-- | Unify empty and None.
+normalizeMintValue :: TxMintValue ViewTx era -> TxMintValue ViewTx era
+normalizeMintValue = \case
+  TxMintValue _ v _ | v == mempty -> TxMintNone
+  other                           -> other
+
+
+-- * Ugrading scripts
+--
+-- The instruction set from V1 may be used as V2,
+-- and we can't determine the language version from the transaction.
+-- When the user uses V1 instructions,
+-- we don't know if they used the V1 tag or the V2 tag.
+-- So it's safe to drop this information.
+
+upgradeScriptInEra :: ScriptInEra era -> ScriptInEra era
+upgradeScriptInEra = \case
+  ScriptInEra SimpleScriptV1InAllegra script ->
+    ScriptInEra SimpleScriptV2InAllegra $ upgradeScript script
+  ScriptInEra SimpleScriptV1InMary script ->
+    ScriptInEra SimpleScriptV2InMary $ upgradeScript script
+  ScriptInEra SimpleScriptV1InAlonzo script ->
+    ScriptInEra SimpleScriptV2InAlonzo $ upgradeScript script
+  other -> other
+
+upgradeScript :: Script SimpleScriptV1 -> Script SimpleScriptV2
+upgradeScript (SimpleScript SimpleScriptV1 script) =
+  SimpleScript SimpleScriptV2 $ upgradeSimpleScript script
+
+upgradeSimpleScript ::
+  SimpleScript SimpleScriptV1 -> SimpleScript SimpleScriptV2
+upgradeSimpleScript = \case
+  RequireSignature hash -> RequireSignature hash
+  RequireAllOf scripts  -> RequireAllOf $ map upgradeSimpleScript scripts
+  RequireAnyOf scripts  -> RequireAnyOf $ map upgradeSimpleScript scripts
+  RequireMOf n scripts  -> RequireMOf n $ map upgradeSimpleScript scripts
+
+
+-- * Review: Change 'TxBodyContent' “phase” to 'ViewTx'
+--
+-- Basically, it's just dropping witness information.
+
+review :: TxBodyContent BuildTx era -> TxBodyContent ViewTx era
+review body =
+  TxBodyContent
+    { txAuxScripts      =                      txAuxScripts     body
+    , txCertificates    = reviewCertificates $ txCertificates   body
+    , txExtraKeyWits    =                      txExtraKeyWits   body
+    , txExtraScriptData = ViewTx
+    , txFee             =                      txFee            body
+    , txIns             = map reviewTxIn     $ txIns            body
+    , txInsCollateral   =                      txInsCollateral  body
+    , txMetadata        =                      txMetadata       body
+    , txMintValue       = reviewMintValue    $ txMintValue      body
+    , txOuts            =                      txOuts           body
+    , txProtocolParams  = ViewTx
+    , txUpdateProposal  =                      txUpdateProposal body
+    , txValidityRange   =                      txValidityRange  body
+    , txWithdrawals     = reviewWithdrawals  $ txWithdrawals    body
+    }
+
+reviewTxIn ::
```

Comment:
Same for all of these `reviewX` functions. We should rename them to something like "toTxInView".